### ISSUE TITLE
[Feature Request] set page HTML

### ISSUE THREAD

**Issue by ejunker** [chrome-php/headless-chromium-php#248]
I'm using this package to generate PDF from Laravel Blade templates. I can easily render the templates to HTML and then I want to generate a PDF. It would be nice if there was an easy built-in way to load HTML and not have to navigate to a URL.

I have found 2 possible ways to do this. The first one is:

```php
$page->navigate('data:text/html,'.rawurlencode($html))->waitForNavigation();
```

The second method mostly works but for some reason the image did not load but that might be a separate issue:

```php
$page->getSession()->sendMessageSync(new Message('Page.setDocumentContent', [
    'frameId' => $page->getFrameManager()->getMainFrame()->getFrameId(),
    'html' => $html,
]));
```

Maybe add a `loadHtml()` or `setHtml()` method which would be much more discoverable. It was not obvious to me how to load HTML until I figured out these tricks.

**Comment by GrahamCampbell** [comment]
Related: https://github.com/chrome-php/chrome/issues/77.

**Comment by ejunker** [comment]
Thanks, I did not see that previous issue #77 

After doing some experimentation it seems like loading from a file is the most reliable.

```php
$htmlPath = tempnam(sys_get_temp_dir(), 'html-').'.html';
file_put_contents($htmlPath, $html);
// waiting for the networkIdle event worked best for me otherwise the images were not loading
$page->navigate('file://'.$htmlPath)->waitForNavigation(Page::NETWORK_IDLE);
```

I also got this method to work which does not use a file.
```php
$page->getSession()->sendMessageSync(new Message('Page.setDocumentContent', [
    'frameId' => $page->getFrameManager()->getMainFrame()->getFrameId(),
    'html' => $html,
]));

// wait for page load event
while (!$page->hasLifecycleEvent(Page::LOAD)) {
    usleep(1);
}
```

I had to add the wait which is not as advanced as `waitForNavigation()` but it works. I experimented with different wait times and surprisingly it didn't seem to make a different and almost always took about 20 iterations through the loop which seems strange.

For what it's worth, I was previously using [knplabs/knp-snappy](https://github.com/KnpLabs/snappy) and they have a [`generateFromHtml()`](https://github.com/KnpLabs/snappy/blob/master/src/Knp/Snappy/AbstractGenerator.php#L222) method that just writes it to a temporary file.

**Comment by enricodias** [comment]
This may be possible to implement using a url created from a blob. I'll take a look into this when I have time.

**Comment by GrahamCampbell** [comment]
> This may be possible to implement using a url created from a blob. I'll take a look into this when I have time.

See #77. There is a limit to how much stuff can be jammed into the url.

**Comment by enricodias** [comment]
> See #77. There is a limit to how much stuff can be jammed into the url.

I did, but a url created using `URL.createObjectURL(blob)` should be able to use up to 500MB. It would create something like `blob:BLOB_URL`. As I understand, the blob itself will be in memory and the url will point to the blob.

**Comment by GrahamCampbell** [comment]
Ah, cool.

**Comment by enricodias** [comment]
I was playing around with the blob idea and found some issues. It does work as I expected, `URL.createObjectURL(blob)` creates a small url that points to the blob in memory. The blob can have up to 500 MB and can be loaded as a normal page. The issue is that all other resources like images and js files will not be loaded in the new page.

- If the resources are on the local disk, chrome will refuse to load them from `file://`. We can teak the security settings to allow this, but I don't like this idea.
- Loading resources with `http` works, but we'll need the full link for each resource, including the original domain. CORS may also hurt us here since we are no longer accessing the original domain.

Check this fork too see the complete code working: https://github.com/enricodias/headless-chromium-php/commit/1c8491d5e75cd6397637d31def8fa3735ebbc26e

I'll take a loot at the `Page.setDocumentContent` solution. If it works and triggers all the important js events, it will probably be a better solution.

**Comment by sanderdlm** [comment]
Is this really a problem this library should solve though?

You can achieve the same result by dumping your generated HTML somewhere inside the public context of your application, navigate to it by URL and delete the HTML file after creation of your PDF. Granted this only works when you're using a router of some sorts, but it is a good solution IMHO.

**Comment by enricodias** [comment]
> You can achieve the same result by dumping your generated HTML somewhere inside the public context of your application, navigate to it by URL and delete the HTML file after

Not in every case. By accessing a file locally you wouldn't be using the same domain as the original website and ajax requests may get blocked due to CORS. I can think about some scenarios where you would need to make changes to the html and keep executing the scripts in the page normally.

You are also assuming that everyone would be using a http server where they have the option to dump a file in a public folder and access it via http. When I started using this package, I was using it in a cli script without any local http server.

And the idea of writing temporary files that may be left behind in case of errors doesn't sounds good either.

**Comment by GrahamCampbell** [comment]
I think a temporary file is fine. Chrome will be making one, anyway. :P

**Comment by enricodias** [comment]
Not stale.

**Comment by msitarzewski** [comment]
I'd love to see this feature as well. I ended up writing the html to a file, then used file:// to load it. Its would be great to bypass that step.

**Comment by GrahamCampbell** [comment]
Implemented in v1.5.0.

**Comment by sanderdlm** [comment]
Cheers, implemented it and works like a charm. No performance gain but removed a lot of overhead from the process of generating PDF (dumping the HTML to a file, navigating to it, cleaning up the temp files, ...). So thanks!
